### PR TITLE
Use Binary Merkle Tree instead of a trie

### PR DESCRIPTION
**Pull request by tomusdrw**
Related: #221 

Also convert uncompressed BEEFY keys to ETH addresses first to reduce gas costs.

### CODE REVIEW

**Comment by tomusdrw** [comment]
CC @philipstanislaus @musnit I think this is now fully matching the JS library. I've added a test case from your test fixture as well.

I'll proceed with changing the implementation in Polkadot now.

**Review by philipstanislaus — APPROVED** [review]
LGTM, thanks for implementing this! 🎉

File: beefy-merkle-root/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+version = "0.1.0"
+authors = ["Parity Technologies <admin@parity.io>"]
+edition = "2018"
+license = "GPL-3.0-or-later WITH Classpath-exception-2.0"
+description = "A no-std/Substrate compatible library to construct binary merkle tree."
+
+[dependencies]
+hex = { version = "0.4", optional = true }
+log = { version = "0.4", optional = true, default-features = false }
+tiny-keccak = { version = "2.0.1", features = ["keccak"], optional = true }
```
Comment (adoerr) [inline beefy-merkle-root/Cargo.toml:12]: Why don't we use version `2.0.2` (the latest) here?

File: beefy-merkle-root/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,807 @@
...
+	/// This vec contains all inner node hashes necessary to reconstruct the root hash given the
+	/// leaf hash.
+	pub proof: Vec<Hash>,
+	/// Number of leaves in the original tree.
+	///
+	/// This is needed to detect a case where we have an odd number of leaves that "get promoted"
+	/// to upper layers.
+	pub number_of_leaves: usize,
+	/// Index of the leaf the proof is for.
+	pub leaf_index: usize,
```
Comment (adoerr) [inline beefy-merkle-root/src/lib.rs:136]: Are those indices `0` or `1` based?

File: beefy-merkle-root/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,807 @@
...
+	let mut leaf = None;
+	let iter = leaves.into_iter().enumerate().map(|(idx, l)| {
+		let hash = H::hash(l.as_ref());
+		if idx == leaf_index {
+			leaf = Some(l);
+		}
+		hash
+	});
+
+	struct ProofCollection {
```
Comment (adoerr) [inline beefy-merkle-root/src/lib.rs:190]: I assume this is the path for a single proof, i.e. all the hashes from leaf `position` to the root. This is not a collection of multiple proofs?

File: beefy-merkle-root/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,807 @@
...
+/// Supported hashing output size.
+///
+/// The size is restricted to 32 bytes to allow for a more optimised implementation.
+pub type Hash = [u8; 32];
+
+/// Generic hasher trait.
+///
+/// Implement the function to support custom way of hashing data.
+/// The implementation must return a [Hash] type, so only 32-byte output hashes are supported.
+pub trait Hasher {
```
Comment (adoerr) [inline beefy-merkle-root/src/lib.rs:45]: Since this trait is public, would it make sense to support conversion to / form, e.g. `sp_core::Hasher`?

**Review by adoerr — APPROVED** [review]
Looks good, some minor remarks.

File: beefy-merkle-root/src/lib.rs (hunk 2 above)
Comment (tomusdrw) [reply beefy-merkle-root/src/lib.rs]: 0-based, added a comment.

File: beefy-merkle-root/src/lib.rs (hunk 3 above)
Comment (tomusdrw) [reply beefy-merkle-root/src/lib.rs]: Yes, we only support single-leaf proof here.

File: beefy-merkle-root/src/lib.rs (hunk 4 above)
Comment (tomusdrw) [reply beefy-merkle-root/src/lib.rs]: I'd rather keep the crate not depending on any substrate stuff.

File: Cargo.toml [hunk 5]
```diff
@@ -8,6 +8,7 @@ members = [
 	"beefy-cli",
 	"beefy-gadget",
 	"beefy-gadget/rpc",
+	"beefy-merkle-root",
```
[the remark this answers is not in the document]
Comment (tomusdrw) [reply Cargo.toml]: I started with `merkle-tree`, but then the main function was only `merkle_root` so I renamed. Later on added proof verification as well, so I think it kind of makes sense to call it `merkle-tree` now :)

**Review by adoerr — APPROVED** [review]

**Pull request merged**